physics/MP/GFDL_2022_v3/gfdl_cld_mp_v3.F90

```diff
@@ -0,0 +1,391 @@
+            pt(i,k)   = gt0(i,kk)
+            w(i,k)    = -vvl(i,kk) * (one+con_fvirt * gq0(i,kk))   &
+                          *  gt0(i,kk) / prsl(i,kk) * (con_rd*onebg)
+            uin(i,k)  = gu0(i,kk)
+            vin(i,k)  = gv0(i,kk)
+            delp(i,k) = del(i,kk)
+            dz(i,k)   = (phii(i,kk)-phii(i,kk+1))*onebg
+            p123(i,k) = prsl(i,kk)
+            qni(i,k)  = 10.
+            q_con(i,k) = 0.0 
+            cappa(i,k) = 0.0 
+         enddo
+      enddo
+
+      ! reset precipitation amounts to zero
+      water0     = 0
+      rain0     = 0
+      ice0      = 0
+      snow0     = 0
+      graupel0  = 0
+ 
+      if(imp_physics == imp_physics_gfdl_v3) then 
+
+        last_step = .false.
+        do_inline_mp = .false. 
+        hs = oro(:) * con_g 
+        gsize = sqrt(garea(:)) 
+
+        call  module_gfdl_cld_mp_driver( qv1, ql1, qr1, qi1, qs1, qg1, qa1, qnl, qni, pt, w,&
+                  uin, vin, dz, delp, gsize, dtp, hs, water0, rain0,                        &
+                  ice0, snow0, graupel0, hydrostatic, iis, iie, kks, kke, q_con, cappa,     &
+                  fast_mp_consv, adj_vmr, te, dte, prefluxw, prefluxr, prefluxi, prefluxs,  &
+                  prefluxg, last_step, do_inline_mp ) 
+
+      else 
+        write(errmsg,'(*(a))') 'Invalid imp_physics option for GFDL MP v3'
+        errflg = 1
+        return
+      endif 
+      tem   = dtp*con_p001/con_day
+
+      ! fix negative values
+      do i = 1, im
+        !rain0(i)    = max(con_d00, rain0(i))
+        !snow0(i)    = max(con_d00, snow0(i))
+        !ice0(i)     = max(con_d00, ice0(i))
+        !graupel0(i) = max(con_d00, graupel0(i))
+        if(water0(i)*tem < rainmin) then
+          water0(i) = 0.0
+        endif
+        if(rain0(i)*tem < rainmin) then
+          rain0(i) = 0.0
+        endif
+        if(ice0(i)*tem < rainmin) then
+          ice0(i) = 0.0
+        endif
+        if(snow0(i)*tem < rainmin) then
+          snow0(i) = 0.0
+        endif
+        if(graupel0(i)*tem < rainmin) then
+          graupel0(i) = 0.0
+        endif
+      enddo
+
+      ! calculate fraction of frozen precipitation using unscaled
+      ! values of rain0, ice0, snow0, graupel0 (for bit-for-bit)
+      do i=1,im
+        prcp0(i) = (rain0(i)+snow0(i)+ice0(i)+graupel0(i)) * tem
+        if ( prcp0(i) > rainmin ) then
+          sr(i) = (snow0(i) + ice0(i)  + graupel0(i)) &
+                      / (rain0(i) + snow0(i) + ice0(i) + graupel0(i))
+        else
+          sr(i) = 0.0
+        endif
+      enddo
+
+      ! convert rain0, ice0, snow0, graupel0 from mm per day to m per physics timestep
+      water0    = water0*tem
+      rain0    = rain0*tem
+      ice0     = ice0*tem
+      snow0    = snow0*tem
+      graupel0 = graupel0*tem
+
+      ! flip vertical coordinate back
+      do k=1,levs
+        kk = levs-k+1
+        do i=1,im
+
+          if (imp_physics == imp_physics_gfdl_v3) then 
```

Comment:
Why not just do this check once at the beginning of the code? Right now it's performed three times within the scheme. It would be best to return an error before beginning computations.